File: gallery/src/main/java/net/faithgen/gallery/adapters/SliderAdapter.java
package net.faithgen.gallery.adapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.smarteist.autoimageslider.SliderViewAdapter;
import com.squareup.picasso.Picasso;

import net.faithgen.gallery.R;
import net.faithgen.gallery.models.Image;
import net.faithgen.gallery.viewholders.SliderHolder;

import java.util.List;

public class SliderAdapter extends SliderViewAdapter<SliderHolder> {

    private Context context;
    private List<Image> images;
    private ImageListener imageListener;
    private LayoutInflater inflater;

    public SliderAdapter(Context context, List<Image> images, ImageListener imageListener) {
        this.context = context;
        this.images = images;
        this.imageListener = imageListener;
        this.inflater = LayoutInflater.from(this.context);
    }

    @Override
    public SliderHolder onCreateViewHolder(ViewGroup parent) {
        return new SliderHolder(inflater.inflate(R.layout.list_item_grid_image, parent, false));
    }

    @Override
    public void onBindViewHolder(SliderHolder viewHolder, int position) {
        Picasso.get()
                .load(images.get(position).getAvatar().getOriginal())
                .error(R.drawable.image_placeholder)
                .placeholder(R.drawable.image_placeholder)
                .into(viewHolder.getImageView());

        viewHolder.getFloatingActionButton().setOnClickListener(v -> {
            if(imageListener != null) imageListener.onImageClicked(images.get(position));
        });
    }

    @Override
    public int getCount() {
        return images.size();
    }

    public interface ImageListener{
        void onImageClicked(Image image);
    }
}

File: gallery/src/main/java/net/faithgen/gallery/adapters/SliderAdapter.kt
package net.faithgen.gallery.adapters

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.smarteist.autoimageslider.SliderViewAdapter
import com.squareup.picasso.Picasso
import net.faithgen.gallery.R
import net.faithgen.gallery.models.Image
import net.faithgen.gallery.viewholders.SliderHolder

/**
 * Adapter to render slider images
 */
final class SliderAdapter(private val context: Context, private val images: List<Image>, private val imageListener: ImageListener?) : SliderViewAdapter<SliderHolder>() {
    private val layoutInflater: LayoutInflater by lazy { LayoutInflater.from(context) }

    override fun onCreateViewHolder(parent: ViewGroup?): SliderHolder {
        return SliderHolder(layoutInflater.inflate(R.layout.list_item_grid_image, parent, false))
    }

    override fun getCount() = images.size

    override fun onBindViewHolder(viewHolder: SliderHolder?, position: Int) {
        Picasso.get()
                .load(images[position].avatar.original)
                .error(R.drawable.image_placeholder)
                .placeholder(R.drawable.image_placeholder)
                .into(viewHolder!!.imageView)

        viewHolder.floatingActionButton.setOnClickListener { v: View? -> imageListener?.onImageClicked(images[position]) }
    }

    interface ImageListener {
        fun onImageClicked(image: Image?)
    }
}